core/app/models/spree/promotion_code/batch_builder.rb

```diff
@@ -2,14 +2,30 @@ class ::Spree::PromotionCode::BatchBuilder
   attr_reader :promotion_code_batch
   delegate :promotion, :number_of_codes, :base_code, to: :promotion_code_batch
 
-  class_attribute :random_code_length, :batch_size, :sample_characters, :join_characters
-  self.random_code_length = 6
-  self.batch_size = 1_000
-  self.sample_characters = ('a'..'z').to_a + (2..9).to_a.map(&:to_s)
-  self.join_characters = "_"
+  DEFAULT_OPTIONS = {
+    random_code_length: 6,
+    batch_size: 1000,
+    sample_characters: ('a'..'z').to_a + (2..9).to_a.map(&:to_s),
+    join_characters: "_"
+  }
 
-  def initialize(promotion_code_batch)
+  [:random_code_length, :batch_size, :sample_characters, :join_characters].each do |attr|
+    define_singleton_method(attr) do
+      Spree::Deprecation.warn "#{name}.#{attr} is deprecated. Use #{name}::DEFAULT_OPTIONS[:#{attr}] instead"
+      DEFAULT_OPTIONS[attr]
+    end
+
+    define_singleton_method(:"#{attr}=") do |val|
+      Spree::Deprecation.warn "#{name}.#{attr}= is deprecated. Use #{name}::DEFAULT_OPTIONS[:#{attr}]= instead"
+      DEFAULT_OPTIONS[attr] = val
+    end
+
+    delegate attr, to: self
+  end
+
+  def initialize(promotion_code_batch, options = {})
     @promotion_code_batch = promotion_code_batch
+    @options = DEFAULT_OPTIONS.merge(options)
```

Comment:
Not necessary, but could we use a attr_reader for @options? Consistency

Reply:
Agree with @tvdeyen , `attr_reader` for `@options` will make this class more consistent.